crates/ef-testing/src/evm_sequencer/evm_state.rs

```diff
@@ -142,12 +143,48 @@ impl EvmState for KakarotSequencer {
         storage.append(allowance_storage);
 
         // Write all the storage vars to the sequencer state.
-        let starknet_address = starknet_address.try_into()?;
         for (k, v) in storage {
-            (&mut self.0.state).set_storage_at(starknet_address, k, v);
+            (&mut self.0.state).set_storage_at(*FEE_TOKEN_ADDRESS, k, v);
         }
         Ok(())
     }
+
+    fn get_storage_at(&mut self, evm_address: &Address, key: U256) -> StateResult<U256> {
+        let keys = split_u256(key).map(Into::into);
+        let keys = get_uint256_storage_var_addresses("storage_", &keys).unwrap(); // safe unwrap: all vars are ASCII
+
+        let starknet_address = compute_starknet_address(evm_address);
+
+        let low = (&mut self.0.state).get_storage_at(starknet_address.try_into()?, keys.0)?;
+        let high = (&mut self.0.state).get_storage_at(starknet_address.try_into()?, keys.1)?;
+
+        let low = U256::from_be_bytes(Into::<FieldElement>::into(low).to_bytes_be());
+        let high = U256::from_be_bytes(Into::<FieldElement>::into(high).to_bytes_be());
+
+        Ok(high << 128 | low)
+    }
+
+    fn get_nonce(&mut self, evm_address: &Address) -> StateResult<U256> {
+        let starknet_address = compute_starknet_address(evm_address);
+
+        let nonce_protocol = (&mut self.0.state)
+            .get_nonce_at(starknet_address.try_into()?)?
+            .0;
+
+        let key = get_storage_var_address("nonce", &[])?;
+        let nonce_managed =
+            (&mut self.0.state).get_storage_at(starknet_address.try_into()?, key)?;
+
+        let nonce = if nonce_managed > nonce_protocol {
```

Comment:
is this safe?
are we sure `nonce_managed > nonce_protocol` <=> account is a CA
and `nonce_managed < nonce_protocol` <=> account is EOA 

?

Reply:
So either both nonces are 0, meaning it will just return 0, either one of the nonce was incremented, meaning this is the used nonce. I can return an error if both nonce are different from 0, would that work?

Reply:
I think then we need to document this "proxy" method a bit, to make sure it's readable

-> when a CA is created with CREATE opcode, its protocol nonce is 0 and its managed nonce becomes 1-> this technique works
-> when an EOA is created its nonce is 0, and its managed nonce will be 0 -> this technique works 
-> when an EOA interacts with the chain, its nonce is incremented, and its managed nonce remains 0 -> this technique works
-> when a CA creates more accts, its managed nonce is incremented -> this technique works